### PR TITLE
Optimize

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,11 +20,14 @@
...
-    "async": "^1.5.2"
+    "async": "^2.6.0"
+  },
+  "optionalDependencies": {
+    "bedrock-redis": "^3.0.0"
   },
   "peerDependencies": {
     "bedrock": "^1.4.1",
-    "bedrock-mongodb": "^3.0.0",
+    "bedrock-mongodb": "^4.0.0",
```
Comment (dlongley) [inline package.json]: May as well go to `5.0.0` here?

Comment (dlongley) [reply package.json]: I wonder if we should do a range for `bedrock-mongodb` ... like `"^3.0.0 - ^5.0.0"` ... is that's the right syntax.

**Review by dlongley — APPROVED** [review]
LGTM -- left some comments on bedrock-mongodb range, we may want to adopt that approach for libs that are compatible with various major versions.

This should be a minor release since redis cache is optional (don't need to go major).

File: lib/index.js [hunk 2]
```diff
@@ -76,20 +76,16 @@ bedrock.events.on('bedrock-mongodb.ready', function init(callback) {
...
  * @return the PublicKey ID.
  */
-api.createPublicKeyId = function(name) {
-  return util.format('%s%s/%s',
-    bedrock.config.key.baseUri,
-    bedrock.config.key.basePath,
-    encodeURIComponent(name));
-};
+api.createPublicKeyId = name => util.format(
+  '%s%s/%s', config.key.baseUri, config.key.basePath, encodeURIComponent(name));
```
Comment (davidlehn) [inline lib/index.js:80]: Could use ES6 syntax:
```js
api.createPublicKeyId = name => `${config.key.baseUri}${config.key.basePath}/${encodeURIComponent(name)}`;
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,16 @@
...
+# 4.1.0 - 2018-02-24
+
+### Changed
+- Use ES6 Syntax.
+- Update bedrock-mongodb dependency.
+- Upgrade to Async@2.
+
+### Added
+- Optional caching of public keys. The current caching implementation is
+  aimed at accelerating http-signature authentication `bedrock-passport`.
```
Comment (davidlehn) [inline CHANGELOG.md:12]: Missing "in" after authentication?

File: lib/index.js [hunk 4]
```diff
@@ -331,17 +367,15 @@ api.getPublicKeys = function(id, actor, options, callback) {
  * @param callback(err) called once the operation completes.
  */
 // FIXME: sync private key label to public key label
-api.updatePublicKey = function(actor, publicKey, callback) {
+api.updatePublicKey = (actor, publicKey, callback) => {
```
Comment (davidlehn) [inline lib/index.js:370]: Should this update the cache too?

File: lib/index.js [hunk 5]
```diff
@@ -3,35 +3,40 @@
...
+const util = require('util');
+const BedrockError = bedrock.util.BedrockError;
 
 // load config defaults
 require('./config');
 
+let cache;
+const {enable: cacheEnabled} = config.key.cache;
+if(cacheEnabled) {
+  cache = require('bedrock-redis');
```
Comment (davidlehn) [inline lib/index.js:22]: since bedrock-redis is in "optionalDependencies" perhaps this should be in a try/catch

**Comment by davidlehn** [comment]
I just noticed this was released from the optimize branch vs master.  When going back to master should do it as a regular merge vs a rebase+merge so the version tag stays the same.

**Comment by dlongley** [comment]
> I just noticed this was released from the optimize branch vs master. When going back to master should do it as a regular merge vs a rebase+merge so the version tag stays the same.

Shouldn't be necessary, as this was already based on master. I pulled it in before someone added anything to master that might disrupt that.

**Pull request merged**